src/main/java/address/sync/CloudSimulator.java

```diff
@@ -1,129 +1,750 @@
+    CloudSimulator(boolean shouldSimulateUnreliableNetwork) {
+        rateLimitStatus = new RateLimitStatus(API_QUOTA_PER_HOUR, API_QUOTA_PER_HOUR, getNextResetTime());
+        resetQuotaAndRestartTimer();
+        this.shouldSimulateUnreliableNetwork = shouldSimulateUnreliableNetwork;
+    }
 
-    public CloudSimulator(boolean shouldSimulateUnreliableNetwork) {
-        this.simulateUnreliableNetwork = shouldSimulateUnreliableNetwork;
+    public static ByteArrayInputStream convertToInputStream(Object object) {
+        try {
+            return new ByteArrayInputStream(JsonUtil.toJsonString(object).getBytes());
+        } catch (JsonProcessingException e) {
+            e.printStackTrace();
+            return null;
+        }
     }
 
     /**
-     * Gets the updated data subjected to simulated random behaviors such as random changes,
-     * random delays and random failures. The data is originally obtained from a given file.
-     * The data is possibly modified in each call to this method and is persisted onto the same file.
-     * When failure condition occurs, this returns an empty optional
+     * Attempts to create a person if quota is available
      *
-     * @return optional wrapping the (possibly corrupted) data, or empty if retrieving failed.
+     * Consumes 1 API usage
+     *
+     * @param addressBookName
+     * @param newPerson
+     * @return a response wrapper, containing the added person if successful
      */
-    public Optional<AddressBook> getSimulatedCloudData(File cloudFile) {
-        System.out.println("Simulating cloud data retrieval...");
+    @Override
+    public RawCloudResponse createPerson(String addressBookName, CloudPerson newPerson, String previousETag) {
+        if (shouldSimulateNetworkFailure()) return getNetworkFailedResponse();
+        if (shouldSimulateSlowResponse()) delayRandomAmount();
+
+        if (!hasApiQuotaRemaining()) return getEmptyResponse(HttpURLConnection.HTTP_FORBIDDEN);
+
         try {
-            AddressBook modifiedData;
-            AddressBook data = XmlFileHelper.getDataFromFile(cloudFile);
-            if (!this.simulateUnreliableNetwork) {
-                return Optional.of(data);
-            }
+            CloudAddressBook fileData = readCloudAddressBookFromFile(addressBookName);
+            CloudPerson returnedPerson = addPerson(fileData.getAllPersons(), newPerson);
+            writeCloudAddressBookToFile(addressBookName, fileData);
 
-            // no data could be retrieved
-            if (RANDOM_GENERATOR.nextDouble() <= FAILURE_PROBABILITY) {
-                System.out.println("Cloud simulator: failure occurred! Could not retrieve data");
-                return Optional.empty();
-            }
+            modifyCloudPersonBasedOnChance(returnedPerson);
+            InputStream bodyStream = convertToInputStream(returnedPerson);
+            String ETag = RawCloudResponse.getETag(bodyStream, false);
```

Comment:
our naming convention would be `eTag` instead of `ETag`, unless there is something before `E`, like `previousETag`.